snakemake/ioutils/evaluate.py

```
@@ -0,0 +1,18 @@
+    {wildcardname} with the wildcard value represented as a string."""
+
+    def inner(wildcards):
+        formatted = expr.format(**{w: repr(v) for w, v in wildcards.items()})
+        try:
+            return eval(formatted, globals())
+        except Exception as e:
+            raise WorkflowError(
+                f"Failed to evaluate expression {expr} with wildcards {wildcards}. "
+                f"Formatted expression: {formatted}",
+                e,
+            )
```

Ruff (0.8.0)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)